pkg/gardenlet/controller/backupbucket/reconciler.go

```diff
@@ -0,0 +1,503 @@
+	extensionBackupBucket *extensionsv1alpha1.BackupBucket,
+) (
+	reconcile.Result,
+	error,
+) {
+	if !sets.NewString(backupBucket.Finalizers...).Has(gardencorev1beta1.GardenerName) {
+		return reconcile.Result{}, nil
+	}
+
+	operationType := v1beta1helper.ComputeOperationType(backupBucket.ObjectMeta, backupBucket.Status.LastOperation)
+	if updateErr := r.updateBackupBucketStatusOperationStart(ctx, backupBucket, operationType); updateErr != nil {
+		return reconcile.Result{}, fmt.Errorf("could not update status after deletion start: %w", updateErr)
+	}
+
+	backupEntryList := &gardencorev1beta1.BackupEntryList{}
+	if err := r.GardenClient.List(ctx, backupEntryList, client.MatchingFields{core.BackupEntryBucketName: backupBucket.Name}); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	associatedBackupEntries := make([]string, 0)
+	for _, entry := range backupEntryList.Items {
+		associatedBackupEntries = append(associatedBackupEntries, client.ObjectKeyFromObject(&entry).String())
+	}
+
+	if len(associatedBackupEntries) != 0 {
+		log.Info("Cannot delete because BackupEntries are still referencing the bucket", "backupEntryNames", associatedBackupEntries)
+		r.Recorder.Eventf(backupBucket, corev1.EventTypeNormal, v1beta1constants.EventResourceReferenced, "cannot delete BackupBucket because the following BackupEntries are still referencing it: %+v", associatedBackupEntries)
+		return reconcile.Result{}, fmt.Errorf("BackupBucket %s still has references", backupBucket.Name)
+	}
+
+	log.Info("No BackupEntries are referencing this BackupBucket, accepting deletion")
+
+	if err := r.deleteGeneratedBackupBucketSecretInGarden(ctx, log, backupBucket); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := r.reconcileExtensionBackupBucketSecret(ctx, backupBucket); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := extensions.DeleteExtensionObject(ctx, r.SeedClient, extensionBackupBucket); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := r.SeedClient.Get(ctx, client.ObjectKeyFromObject(extensionBackupBucket), extensionBackupBucket); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return reconcile.Result{}, err
+		}
+	} else {
+		if lastError := extensionBackupBucket.Status.LastError; lastError != nil {
+			r.Recorder.Event(backupBucket, corev1.EventTypeWarning, gardencorev1beta1.EventDeleteError, lastError.Description)
+
+			if updateErr := r.updateBackupBucketStatusError(ctx, backupBucket, lastError.Description+" Operation will be retried.", lastError); updateErr != nil {
+				return reconcile.Result{}, fmt.Errorf("could not update status after deletion error: %w", updateErr)
+			}
+			return reconcile.Result{}, errors.New(lastError.Description)
+		}
+		log.Info("Extension BackupBucket not yet deleted", "extensionBackupBucket", client.ObjectKeyFromObject(extensionBackupBucket))
+		return reconcile.Result{RequeueAfter: RequeueDurationWhenResourceDeletionStillPresent}, nil
```

Comment:
Have you thought about making the `ExtensionStatusChanged` also return true for delete events? Then you would not need to requeue here based on a timer.

Reply:
Make sense, it works.

Reply:
But it flakes when I stress tested. So I changed it back.

Reply:
Alright let's keep it like this for now.